### PR TITLE
Day 2 Part 1 - Specs and more blocks this time around

### PR DESCRIPTION
**Pull request by coopergillan**
### What

Add solution for Day 2 Part 1 with specs and more blocks this time around. Trying to build off some of the progress from Day 1 over in #1.

### Why

It's fun to learn a new programming language, especially one that you use in your job!

@raymondberg - thanks for your help before - check it out only if you get a minute
@misterjacko might be curious, too

### CODE REVIEW

File: day2/lib/day2.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    def initialize(password_list)
+      @password_list = password_list
+    end
+
+    def self.from_file(filepath)
+      new(File.foreach(filepath).map { |line| line.chomp })
+    end
+
+    def count_valid_entries(entry_klass)
+      @password_list.map { |raw_entry| entry_klass.new(raw_entry).valid? }.count(true)
```
Comment (raymondberg) [inline day2/lib/day2.rb:14]: Look into `.reduce` here

File: day2/lib/day2.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    end
+  end
+
+  class Part1
+    class Entry
+      attr_reader :min_chars, :max_chars, :char, :password
+
+      def initialize(raw_entry)
+        char_limits, raw_char, @password = raw_entry.split(" ")
+        @min_chars, @max_chars = char_limits.split("-").map { |raw_char| raw_char.to_i }
```
Comment (raymondberg) [inline day2/lib/day2.rb:24]:
```suggestion
        @min_chars, @max_chars = char_limits.split("-").map(&:to_i)
```

File: day2/lib/day2.rb (hunk 1 above)
Comment (coopergillan) [reply day2/lib/day2.rb:14]: This is what I've ended up with. I can't decide if I like it more or not. Either this:

```ruby
def count_valid_entries(entry_klass)
  @password_list.map do |raw_entry|
    entry_klass.new(raw_entry)
  end.reduce(0) { |count, value| count + (value.valid? ? 1 : 0) }
end
```

Or this:

```ruby
def count_valid_entries(entry_klass)
  @password_list.map do |raw_entry|
    entry_klass.new(raw_entry)
  end.reduce(0) do |count, value|
    count + (value.valid? ? 1 : 0)
  end
end
```

Comment (coopergillan) [reply day2/lib/day2.rb:14]: Ah, here we go. I think I like this the best:

```ruby
def count_valid_entries(entry_klass)
  @password_list.reduce(0) do |valid_entries, raw_entry|
    valid_entries + (entry_klass.new(raw_entry).valid? ? 1 : 0)
  end
end
```

File: day2/lib/day2.rb (hunk 2 above)
Comment (coopergillan) [reply day2/lib/day2.rb:24]: Huge fan. Applied it here and two other places via 78ff478528788bfba80dfc8d8d87a6413e0faca3.

**Pull request merged**